### PR TITLE
Added new metric to Rados Bench - CPU cycles per operation and Fixing the cbt config file generator not including monitoring profiles

### PR DESCRIPTION
**Pull request by AmnonHanuhov**
I made 3 main changes:
1. Added a metric for Rados Bench to be able to present the number of cpu cycles per operation
2. Fixed an issue where setting.py wasn't preparing the cbt config file correctly (Monitoring profiles were not added)
3. Cleaned the name-space in radosbench.py by giving more accurate names to some of the variables

### CODE REVIEW

**Comment by AmnonHanuhov** [comment]
@rzarzynski @ronen-fr Tagging you here for the review. Couldn't officially assign you as reviewers because only the owner of the repo can do that.

File: benchmark/radosbench.py [hunk 1]
```diff
@@ -244,17 +244,38 @@ def analyze(self, out_dir):
...
 
-    def _compare_client_results(self, client_run, fnames):
+    def get_total_ops(self, json_output):
+        json_str = json.dumps(json_output) #json_str is the rados bench output as string
+        total_ops_type = re.search(r'writes', json_str, re.M|re.I)
+        if total_ops_type is not None:
+            total_ops_key = "Total writes made"
+        else:
+            total_ops_key = "Total reads made"
+        print (json_output[total_ops_key.replace(',' , '')])
```
Comment (ronen-fr) [inline benchmark/radosbench.py]: Can this be computed once?

File: benchmark/radosbench.py [hunk 2]
```diff
@@ -278,18 +299,21 @@ def evaluate(self, baseline):
         for run in runs:
             out_dirs = [os.path.join(self.out_dir, run),
                         os.path.join(baseline.out_dir, run)]
+            for d in out_dirs:
+                print (d)
```
Comment (ronen-fr) [inline benchmark/radosbench.py]: Is this part of the final code?

File: setup.sh [hunk 3]
```diff
@@ -11,14 +11,14 @@ sudo yum install -y psmisc util-linux coreutils xfsprogs e2fsprogs findutils \
 
 MIRROR="http://mirror.hmc.edu/fedora/linux/releases/22/Everything/x86_64/os/Packages"
 
-wget ${MIRROR}/p/pdsh-2.31-3.fc22.x86_64.rpm
-wget ${MIRROR}/p/pdsh-2.31-3.fc22.x86_64.rpm
-wget ${MIRROR}/p/pdsh-rcmd-ssh-2.31-3.fc22.x86_64.rpm
-wget ${MIRROR}/c/collectl-4.0.0-1.fc22.noarch.rpm
-wget ${MIRROR}/i/iftop-1.0-0.9.pre4.fc22.x86_64.rpm
-wget ${MIRROR}/i/iperf3-3.0.10-1.fc22.x86_64.rpm
+#wget ${MIRROR}/p/pdsh-2.31-3.fc22.x86_64.rpm
```
Comment (ronen-fr) [inline setup.sh]: what are these removals related to?

File: benchmark/radosbench.py (hunk 2 above)
Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: The printing was for debugging purposes, got rid of it in the new commit.

File: setup.sh (hunk 3 above)
Comment (AmnonHanuhov) [reply setup.sh]: This mirror did not seem to work, checked again now and it still doesn't work. In the new commit I updates the script with a new and working mirror.

File: benchmark/radosbench.py [hunk 4]
```diff
@@ -244,17 +244,28 @@ def analyze(self, out_dir):
...
-    def _compare_client_results(self, client_run, fnames):
+    def _compare_client_results(self, client_run, fname_from_evaluate, out_dirs):
         # normalize the names
         aliases = {'bandwidth': 'Bandwidth (MB/sec)',
                    'iops_avg': 'Average IOPS',
                    'iops_stddev': 'Stddev IOPS',
                    'latency_avg': 'Average Latency(s)'}
         json_outputs = []
         compare_results = []
+
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: Got rid of them in the new commit.

File: benchmark/radosbench.py [hunk 5]
```diff
@@ -244,17 +244,28 @@ def analyze(self, out_dir):
...
+        
+
         for fname in fnames:
             with open(fname) as f:
                 json_outputs.append(json.load(f))
+                
+        i = 0
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: Switched counter management to enumerate as requested in the new commit.

File: monitoring.py [hunk 6]
```diff
@@ -86,6 +90,17 @@ def stop(self, directory):
             common.pdsh(self.nodes, 'sudo chown {user}.{user} {dir}/perf/perf_stat.*'.format(
                     user=self.user, dir=directory))
 
+    def get_cpu_cycles(out_dir):
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply monitoring.py]: Moved it inside the function scope as requested in the new commit.

File: benchmark/radosbench.py [hunk 7]
```diff
@@ -289,27 +280,21 @@ def evaluate(self, baseline):
...
-                    compare_results = self._compare_client_results(client_run, fname, out_dirs)
+                    results_files_paths  = [os.path.join(d, fname) for d in out_dirs]
+                    compare_results = self._compare_client_results(client_run, results_files_paths)
                     rejected = sum(not result.accepted for result in compare_results)
                     results.extend(compare_results)
             # TODO: check results from monitors
         return results
 
-    #getting the number of cpu cycles calculated by Perf Stat
-    def get_cpu_cycles(out_dir):
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: Fixed the commit conflicts

File: monitoring.py [hunk 8]
```diff
@@ -138,3 +153,13 @@ def monitor(directory):
     yield
     for m in monitors:
         m.stop(directory)
+
+def get_cpu_cycles(out_dir):
+    #check if perf stat is configured
+    for monitoring_profile in _get_all():
+        if(isinstance(monitoring_profile, PerfMonitoring):
+            return monitoring_profile.get_cpu_cycles() #if it is, then return the number of cycles
+    #else
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply monitoring.py]: Added the error handling in the last commit

File: benchmark/radosbench.py (hunk 1 above)
Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: Yes, I changed the logic in the latest commit (fc3670d)

File: benchmark/radosbench.py [hunk 9]
```diff
@@ -284,12 +295,21 @@ def evaluate(self, baseline):
...
                     results.extend(compare_results)
             # TODO: check results from monitors
         return results
 
+    #getting the number of cpu cycles calculated by Perf Stat
+    def get_cpu_cycles(out_dir):
+        return monitoring.get_cpu_cycles(out_dir)
+
+    def get_total_ops(json_output):
+        json_str = json.dumps(json_output) #json_str is the rados bench output as string
```
[the remark this answers is not in the document]
Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: You are right, it was a bit of an overkill, I wrote a better solution to the problem in the latest commit (fc3670d)

File: benchmark/radosbench.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rzarzynski) [reply benchmark/radosbench.py]: OK, the commit for `enumerate` looks fine.

File: setup.sh [hunk 10]
```diff
@@ -9,16 +9,15 @@ sudo yum install -y psmisc util-linux coreutils xfsprogs e2fsprogs findutils \
   git wget bzip2 make automake gcc gcc-c++ kernel-devel perf blktrace lsof \
   redhat-lsb sysstat screen python-yaml ipmitool dstat zlib-devel ntp
 
-MIRROR="http://mirror.hmc.edu/fedora/linux/releases/22/Everything/x86_64/os/Packages"
+MIRROR="http://rpm.pbone.net/index.php3/stat/4"
```
Comment (rzarzynski) [inline setup.sh]: I think this commit deserves a dedicated PR.

File: monitoring.py [hunk 11]
```diff
@@ -91,6 +90,7 @@ def stop(self, directory):
                     user=self.user, dir=directory))
 
     def get_cpu_cycles(self, out_dir):
+        import re
```
Comment (rzarzynski) [inline monitoring.py]: :+1:

File: benchmark/radosbench.py [hunk 12]
```diff
@@ -255,22 +255,31 @@ def get_total_ops(self, json_output):
 
     #getting the number of cpu cycles calculated by Perf Stat
     def get_cpu_cycles(self, perf_out_dir):
-        return monitoring.get_cpu_cycles(perf_out_dir)
+        try:
+            num_cpu_cycles = monitoring.get_cpu_cycles(perf_out_dir)
```
Comment (rzarzynski) [inline benchmark/radosbench.py]: Is `num_cpu_cycles` used anywhere?

File: benchmark/radosbench.py [hunk 13]
```diff
@@ -245,13 +245,9 @@ def analyze(self, out_dir):
...
 
     def get_total_ops(self, json_output):
-        json_str = json.dumps(json_output) #json_str is the rados bench output as string
-        total_ops_type = re.search(r'writes', json_str, re.M|re.I)
-        if total_ops_type is not None:
-            total_ops_key = "Total writes made"
-        else:
-            total_ops_key = "Total reads made"
-        return json_output[total_ops_key.replace(',' , '')]
+        search_key = "made"
```
Comment (rzarzynski) [inline benchmark/radosbench.py]: Might be useful to add a comment exemplifying what we're looking for (`Total writes made`, `Total reads made`).

Comment (AmnonHanuhov) [reply benchmark/radosbench.py]: You are right. I added the comment like you suggested in the latest commit.

**Comment by tchaikov** [comment]
@AmnonHanuhov could you remove the merge commit from this PR?

File: benchmark/radosbench.py [hunk 14]
```diff
@@ -244,17 +244,28 @@ def analyze(self, out_dir):
...
         # normalize the names
         aliases = {'bandwidth': 'Bandwidth (MB/sec)',
                    'iops_avg': 'Average IOPS',
                    'iops_stddev': 'Stddev IOPS',
                    'latency_avg': 'Average Latency(s)'}
         json_outputs = []
         compare_results = []
+
+        
+
```
Comment (tchaikov) [inline benchmark/radosbench.py]: still too many empty lines.

File: benchmark/radosbench.py [hunk 15]
```diff
@@ -244,17 +244,28 @@ def analyze(self, out_dir):
...
         for fname in fnames:
             with open(fname) as f:
                 json_outputs.append(json.load(f))
+                
+        i = 0
+        for fpath in out_dirs:
+            total_ops = get_total_ops(json_outputs[i])
+            json_outputs[i]['Cycles per operation'] = get_cpu_cycles(fpath) / int(total_ops)
+            i = i+1
```
Comment (tchaikov) [inline benchmark/radosbench.py]: nit,  could put
```py
i += 1
```

File: benchmark/radosbench.py [hunk 16]
```diff
@@ -244,17 +244,28 @@ def analyze(self, out_dir):
...
+
+
```
Comment (tchaikov) [inline benchmark/radosbench.py]: could you remove the extra empty line?

File: setup.sh (hunk 3 above)
Comment (tchaikov) [inline setup.sh]: could you move this change into a separated commit?

Comment (AmnonHanuhov) [reply setup.sh]: I think I should also move this into a seperated PR

File: common.py [hunk 17]
```diff
@@ -108,7 +108,7 @@ def get_localnode(nodes):
 
     # if more than one node is listed, fallback to pdsh
     nodes_list = expanded_node_list(nodes);
-    if len(nodes) > 1:
+    if len(nodes_list) > 1:
```
Comment (rzarzynski) [inline common.py]: That's an artifact from the development process, I guess. ;-)
The fix for this issue has been already merged.

**Comment by rzarzynski** [comment]
Can't find a way to comment on a file with unmodified content in GitHub, so posting here:

`setup.sh` got change of permissions: `a+x` – from `0644` to `0755`. If we want to keep this change, then a dedicated commit would be strongly advised.

**Comment by markhpc** [comment]
Needs rebase now

**Comment by rzarzynski** [comment]
I think the changes were included in #211 and we can close this PR now.
